Add Google Drive and AWS S3 as a Remote Storage option

cs7-shrey opened the pull request:
## Description
This pull requests is a first version of Google Drive and AWS S3 integration in the datashuttle API and TUI.

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**

**What does this PR do?**
This PR is to discuss the approach for implementing Google Drive and AWS S3 as a remote storage option.  

## References
Issue #407 

## How has this PR been tested?
This PR was manually tested in the Datashuttle's Python API and TUI. Comprehensive tests are yet to be written.

## Is this a breaking change?
No

## Does this PR require an update to the documentation?
Yes

## Checklist:

- [x] The code has been tested locally
- [ ] Tests have been added to cover all new functionality
- [ ] The documentation has been updated to reflect any changes
- [ ] The code has been formatted with [pre-commit](https://pre-commit.com/)

cs7-shrey commented:
Hey @adamltyson @JoeZiminski, I know reviewing this PR wouldn't be feasible at the moment. But it would be great if you can try it out locally whenever you have time and let me know if it works correctly / any suggestions you'd have. No hurry at all, I understand if it takes a bit longer.

JoeZiminski commented on datashuttle/utils/rclone.py:
```diff
@@ -108,6 +108,64 @@ def setup_rclone_config_for_ssh(
+    )
+    client_secret_key_value = (
+        f"client_secret {cfg['gdrive_client_secret']} "
+        if cfg.get("gdrive_client_secret", None)
```
Also, we could then simplify these lines only to:

```
client_secret_key_value = cfg["gdrive_client_secret"] or ""
```

JoeZiminski commented on datashuttle/utils/gdrive.py:
```diff
@@ -0,0 +1,74 @@
+        if cfg.get("gdrive_client_secret", None)
+        else ""
+    )
+    output = rclone.call_rclone(
```
Some of these arguments to the rclone call are a little mysterious, a short comment above would be useful just to explain to readers what is going on. Of course, its always possible to check the rclone docs, but in general it is more efficient to leave a short comment to give the pertinent details.

JoeZiminski commented on datashuttle/utils/gdrive.py:
```diff
@@ -0,0 +1,74 @@
+        pipe_std=True,
+    )
+
+    # TODO: make this more robust
```
TODO here

cs7-shrey replied:
Actually, I put this TODO so as to be absolutely sure that rclone would behave in the same sequence of steps as we expect it to as I hackishly put together some rclone commands before. But with more and more testing and reading rclone's docmentation, I think it will most likely behave as expected. 

So, the only concern here is error handling, for the json loading, it will raise any exception itself if the json doesn't have the structure we expect. To be sure that the rclone's output message is what we suppose it should be, shall we check something like 
`if <some_string> not in message: raise Exception`  ?

what do you suggest?

JoeZiminski commented on datashuttle/utils/gdrive.py:
```diff
@@ -0,0 +1,74 @@
+        answer = False
+
+    if log:
+        utils.log(message)
```
Can the answer also be logged here?

cs7-shrey commented on datashuttle/configs/canonical_configs.py:
```diff
@@ -39,9 +40,16 @@ def get_canonical_configs() -> dict:
         "central_path": Optional[Union[str, Path]],
-        "connection_method": Optional[Literal["ssh", "local_filesystem"]],
+        "connection_method": Optional[
+            Literal["ssh", "local_filesystem", "gdrive", "aws_s3"]
```
I just thought, it would be super explicit and clear to the developers if we explicitly mentioned "s3" with aws. Just "aws" can be sufficient but I think "aws_s3" will be a little more cognitive load to read and understand. Happy to change if you'd like.

JoeZiminski replied:
Yes I'm not sure there is a clear best approach, in general I think users will be more neuroscience than developer background and keeping it as simple as possible would be nice where feasible. I don't actually know anything about AWS terminology😆 I just saw S3 means 'Simple Storage Solutions', I was looking for an S2 and S1 versions haha. I think in this case we can just keep it to `aws` as their other products are not so relevant in this context, so "aws" = "aws storage" = "aws_s3" is implied. I think that makes sense 🤔

JoeZiminski replied:
yes I think so that works well for consistency

cs7-shrey commented on datashuttle/utils/folders.py:
```diff
@@ -537,6 +557,61 @@ def search_for_folders(
     return all_folder_names, all_filenames
 
 
+def search_remote_central_for_folders(
```
I tried it before for ssh too and it did work. We can centralise this for ssh too.

cs7-shrey commented on datashuttle/tui/configs.py:
```diff
@@ -253,7 +314,15 @@ def on_radio_set_changed(self, event: RadioSet.Changed) -> None:
             display_ssh = True if label == "SSH" else False
 
-        self.switch_ssh_widgets_display(display_ssh)
+            if label == "SSH":
```
That's a fantastic idea! Works like a charm.

cs7-shrey commented on datashuttle/utils/aws.py:
```diff
@@ -0,0 +1,50 @@
+
+    if not check_if_aws_bucket_exists(cfg):
+        bucket_name = cfg["central_path"].as_posix().strip("/").split("/")[0]
+        utils.print_message_to_user(
```
Actually, it just shows a warning to the user so that he/she can change their bucket name. Maybe it should be changed to raise error on bucket not found as the transfer can't take place if the bucket does not exist. What do you suggest?

cs7-shrey commented on datashuttle/tui/configs.py:
```diff
@@ -631,4 +883,32 @@ def get_datashuttle_inputs_from_widgets(self) -> Dict:
             None if central_host_username == "" else central_host_username
         )
 
+        # Google Drive specific
```
I have one more suggestion. Do you think we should also "null" out the configs of the previous connection method if the user saves a new connection method?

cs7-shrey replied:
Cool. That's a valid point.

JoeZiminski reviewed (changes requested):
Hi @cs7-shrey this is looking really nice, some small comments but on the implementation side this is basically done, next we can tackle tests and documentation. Thanks for this comprehensive, excellent piece of work!

On the testing side, I will go through and test manually a little more, and then we also have:
1) Internal NIU tests that will actually connect to AWS and Google Drive. This will not be included in this repo and will run separately, maybe once a week. A benefit of these tests is that they test the actual behaviour, but a downside is they do not run on every PR and so cannot quickly catch when a PR accidently breaks existing behaviour.

(Incidentally, I just had a thought on this, maybe we can use Github secrets to handle the testing in the CI! will try this out over the next few days)

2) Tests included in this repository. A downside of these is that we will not be able to actually connect to SWC or Google Drive. However, we can get quite far by mocking relevant functions. 

For 2) we can go through the code and see where there is opportunity for testing. It can be useful to run the commands in code and through the TUI yourself pretending to be a user, and see what the possible failure modes might be. We can also extend existing tests for the new use cases, I'll leave another comment in the next few days with anything further I think of but please feel free to get started. Some things off the top of my head:

- Extending the [TUI configs tests](https://github.com/neuroinformatics-unit/datashuttle/blob/main/tests/tests_tui/test_tui_configs.py) to Google Drive / AWS
- We could mock the low level `call_rclone` function in `setup_rclone_config_for_gdrive` and check the `DataShuttle` class config setup are working. There is an example on how to programmatically trigger the `input` [here](https://github.com/neuroinformatics-unit/datashuttle/blob/44bd759e459be1980fccc7f942b83588bbc9d83b/tests/ssh_test_utils.py#L47)
- quick test that passing bad combinations of the gdrive / aws settings (e.g. one but not both) triggers an error
- there is a lot of functionality in the `aws` and `gdrive` modules. These could be tested via unit test, or implicitly tested by testing the higher level functions (e.g. through the TUI) or `DataShuttle` class set up functions. I'll let you lead on what is best for these, in general I feel high-level integration tests are nice because you end up testing the low level functions as well as how they interact. But, it is worth going through each function and thinking about how to test it, and what might break it. For key functions, it can be also nice to unit test them, barraging them with a wide range of inputs.
- The `search_gdrive_or_aws_for_folders` is really key function, I'll look into this function a little more and see if it can be integrated with SSH and testing combined.

**Outside the scope of this PR**
- for the setup aws, setup gdrive (and setup ssh) TUI screens, it would be really nice to tests these but they are a real pain (which is why SSH is not tested). At some point we will need to test them but this is getting outside the scope of this PR and we can create some issues. 
- At some point we should test all the decorators are working, this is outside the scope of this PR, I'll make an issue.

Also, the test coverage on the package is not 100% so if you see any areas of the codebase that are untested and you think are important please let me know. Maybe we can meet this week to discuss, as there is a lot here. For me, testing is the most boring part 😅 and I appreciate it's a bit of a slog, especially after all the work done on the codebase. It's worth remembering that this code will hopefully be being used 2, 5 even 10 years from now with many many PRs written over the top of them. Most bugs are not introduced when code is written, but much later when new functionality accidentally breaks old functionality. These tests will ensure these features are working as intended for years to come!

For me to do:
- see if `search_gdrive_or_aws_for_folders` can be used for SSH too
- make NIU accounts for AWS and Google Drive and see if we can connect to them through GitHub CI with secrets
- manual test a little more and think about tests to add 
- make an issue about testing decorators and the connection set up screens.

Let me know if anything is not clear, thanks Shrey!

cs7-shrey commented on datashuttle/utils/folders.py:
```diff
@@ -537,6 +552,63 @@ def search_for_folders(
+        f"{cfg.get_rclone_config_name()}:{search_path.as_posix()} "
+        f'--include "{search_prefix}"',
+    )
+    output = subprocess.run(
```
Sure. I ran into some circular import issues previously. Now it's fixed.

cs7-shrey commented on datashuttle/tui/shared/configs_content.py:
```diff
@@ -586,6 +737,98 @@ def fill_widgets_with_project_configs(self) -> None:
+            not connection_method
+            or connection_method == "local_filesystem"
+            or not self.interface
+            or connection_method != self.interface.get_configs()["connection_method"]
```
Do you mean the `connection_method != self.interface.get_configs()["connection_method"]` ?
It's actually the case when the user switches radiobuttons for connection method. If the current radiobutton is not the same as the connection method set in configs, the `setup_connection_button` will not be displayed.

cs7-shrey commented on datashuttle/utils/rclone.py:
```diff
@@ -150,6 +150,91 @@ def setup_rclone_config_for_ssh(
+        )
+
+    if log:
+        log_rclone_config_output()
```
The log output would then also contain the client secret if the user entered one which I don't think would be good.

JoeZiminski replied:
Sorry just realized I never responded to this, agree let's not do that :D

cs7-shrey commented:
Hey @JoeZiminski I refactored the google drive connection to the service account method. Also, the `rclone ls` command was silently exiting even if it did not have access to the google drive folder ID so, I changed the approach used in `check_successful_connection_and_raise_error_on_fail`. It now checks the connection by creating a `temp.txt` file on the remote and deleting it. Although, the error message is throws is a bit longer. Let me know what you think of this.

Also, do you remember about a `# fix` comment I had in `setup_gdrive.py`? In the part of the code that would execute when a "cancel" button was pressed during google drive setup connection. The thing is, if the cancel button is pressed when the final rclone connection setup (via browser) is started (in a worker), the `self.setup_worker.cancel()` will surely cancel the worker but the  `call_rclone` function that invoked a `subprocess.run` is still running. It's terminal equivalent is if you call `rclone config create ...` and the browser opens up but you close the browser tab; the `rclone config create ...` still hasn't exited and you'd need to force stop it by doing something like `ctrl + c`.

Long story short, on clicking "cancel", the worker is cancelled but the subprocess that was started by it (which in turn opened the google's oauth in a browser) is still running. 

The problem is, since the previous rclone command is still running, it is still using port `53682` for the oauth setup. If the user tries to setup connection again it fails with rclone saying something like `port 53682 is already in use` and the the user would have to quit the TUI and run it again.

The easiest fix I think is to kill whatever process is listening on port `53682`. I don't know how good of an approach this is since we're hardcoding the port here.
```py
import psutil
def kill_process_on_port(port):
    for conn in psutil.net_connections(kind='inet'):
        if conn.laddr.port == port and conn.status == psutil.CONN_LISTEN:
            pid = conn.pid
            if pid:
                try:
                    proc = psutil.Process(pid)
                    print(f"Killing PID {pid} ({proc.name()}) listening on port {port}")
                    proc.kill()
                except Exception:
                    pass
```

Another way would be to change the `call_rclone` function itself and maybe use `subprocess.Popen` that can give us more control over managing the process. Let me know what you think.

cs7-shrey commented on datashuttle/tui/interface.py:
```diff
@@ -499,3 +504,72 @@ def setup_key_pair_and_rclone_config(
+                gdrive_client_secret, service_account_filepath
+            )
+            self.google_drive_rclone_setup_process = process
+            self.gdrive_setup_process_killed = False
```
Kind of. Awaiting successful connection google drive connection would throw an error since killing the process gives a non-zero return code. So, this makes sure that an error is not raised if the process was manually killed which i think would be possible in a race condition as you said.

JoeZiminski replied:
Yup exactly in case the process has been killed when this is called: `if process.returncode != 0:`, sounds good!

JoeZiminski commented on datashuttle/tui/interface.py:
```diff
@@ -499,3 +504,72 @@ def setup_key_pair_and_rclone_config(
+
+        process = self.google_drive_rclone_setup_process
+
+        if process.poll() is not None:
```
okay great, unless I am mistaken I think `poll()` counter-intuitively will return `None` when the process is _running_, so the line should be:

```
if process.poll() is None:
    self.gdrive_setup_process_killed = True
    process.kill()
```
